### ISSUE TITLE
Facet Query

### ISSUE THREAD

**Issue by dtosato** [NightOwl888/BoboBrowse.Net#8]
Following [the example here](https://linkedin.jira.com/wiki/display/BOBO/Facet+Query), I found that writing a query with the standard Lucene API and Bobo does not produce exactly the same result.
In particular the score produced by the Lucene query is normalized, while the one produced by the Bobo query is not. Why?
Here a test for you.
```c#
[TestMethod]
public void FacetQueryTest()
{
	// Create index.
	string indexPath = @"c:\temp\FacetQueryTest";
	{
		using (Lucene.Net.Store.Directory directory = FSDirectory.Open(new System.IO.DirectoryInfo(indexPath)))
		{
			int number;
			using (IndexWriter modifier = new IndexWriter(directory, new StandardAnalyzer(Lucene.Net.Util.Version.LUCENE_30), true, Lucene.Net.Index.IndexWriter.MaxFieldLength.UNLIMITED))
			{
				{
					number = 1;
					string numberFormatted = number.ToString("00", System.Globalization.CultureInfo.InvariantCulture);
					Document doc = new Document();
					doc.Add(new Field("number", numberFormatted, Field.Store.YES, Field.Index.NOT_ANALYZED));

					modifier.AddDocument(doc);
				}

				{
					number = 2;
					string numberFormatted = number.ToString("00", System.Globalization.CultureInfo.InvariantCulture);
					Document doc = new Document();
					doc.Add(new Field("number", numberFormatted, Field.Store.YES, Field.Index.NOT_ANALYZED));
					modifier.AddDocument(doc);
				}

				{
					number = 3;
					string numberFormatted = number.ToString("00", System.Globalization.CultureInfo.InvariantCulture);
					Document doc = new Document();
					doc.Add(new Field("number", numberFormatted, Field.Store.YES, Field.Index.NOT_ANALYZED));
					modifier.AddDocument(doc);
				}

				{
					number = 4;
					string numberFormatted = number.ToString("00", System.Globalization.CultureInfo.InvariantCulture);
					Document doc = new Document();
					doc.Add(new Field("number", numberFormatted, Field.Store.YES, Field.Index.NOT_ANALYZED));
					modifier.AddDocument(doc);
				}

				{
					number = 5;
					string numberFormatted = number.ToString("00", System.Globalization.CultureInfo.InvariantCulture);
					Document doc = new Document();
					doc.Add(new Field("number", numberFormatted, Field.Store.YES, Field.Index.NOT_ANALYZED));
					modifier.AddDocument(doc);
				}

				{
					number = 6;
					string numberFormatted = number.ToString("00", System.Globalization.CultureInfo.InvariantCulture);
					Document doc = new Document();
					doc.Add(new Field("number", numberFormatted, Field.Store.YES, Field.Index.NOT_ANALYZED));
					modifier.AddDocument(doc);
				}

				{
					number = 7;
					string numberFormatted = number.ToString("00", System.Globalization.CultureInfo.InvariantCulture);
					Document doc = new Document();
					doc.Add(new Field("number", numberFormatted, Field.Store.YES, Field.Index.NOT_ANALYZED));
					modifier.AddDocument(doc);
				}

				{
					number = 8;
					string numberFormatted = number.ToString("00", System.Globalization.CultureInfo.InvariantCulture);
					Document doc = new Document();
					doc.Add(new Field("number", numberFormatted, Field.Store.YES, Field.Index.NOT_ANALYZED));
					modifier.AddDocument(doc);
				}

				{
					number = 9;
					string numberFormatted = number.ToString("00", System.Globalization.CultureInfo.InvariantCulture);
					Document doc = new Document();
					doc.Add(new Field("number", numberFormatted, Field.Store.YES, Field.Index.NOT_ANALYZED));
					modifier.AddDocument(doc);
				}

				{
					number = 10;
					string numberFormatted = number.ToString("00", System.Globalization.CultureInfo.InvariantCulture);
					Document doc = new Document();
					doc.Add(new Field("number", numberFormatted, Field.Store.YES, Field.Index.NOT_ANALYZED));
					modifier.AddDocument(doc);
				}
			}
		}
	}

	// Field.
	string field = "number";

	// Facet.
	IFacetHandler numberFacetHandler = new SimpleFacetHandler(field);

	ICollection<IFacetHandler> handlerList = new IFacetHandler[] {  
		numberFacetHandler};

	// Lucene index.
	using (Lucene.Net.Store.Directory idx = FSDirectory.Open(new System.IO.DirectoryInfo(indexPath)))
	{
		using (IndexReader reader = IndexReader.Open(idx, true))
		{
			// Bobo reader.
			using (BoboIndexReader boboReader = BoboIndexReader.GetInstance(reader, handlerList))
			{
				string v1 = "01", v2 = "05", v3 = "09";
				float b1 = 1.1f, b2 = 0.9f, b3 = 1.0f;
				int totalHits, totalHits2;
				BrowseResult result, result2;

				// Bobo.
				{
					BrowseRequest browseRequest = new BrowseRequest();
					browseRequest.Count = 10;
					browseRequest.Offset = 0;
					browseRequest.FetchStoredFields = true;

					BrowseSelection sel = new BrowseSelection(field);
					sel.Values = new string[] { v1, v2, v3 };
					browseRequest.AddSelection(sel);

					Dictionary<string, float> boostMap = new Dictionary<string, float>();
					boostMap.Add(v1, b1);
					boostMap.Add(v2, b2);
					boostMap.Add(v3, b3);

					FacetTermQuery resultQ = new FacetTermQuery(sel, boostMap);
					browseRequest.Query = resultQ;

					FacetSpec spec = new FacetSpec();
					spec.MaxCount = 10;
					browseRequest.SetFacetSpec(field, spec);

					IBrowsable browser = new BoboBrowser(boboReader);
					result = browser.Browse(browseRequest);

					totalHits = result.NumHits;
				}

				// Lucene.
				{
					BrowseRequest browseRequest = new BrowseRequest();
					browseRequest.Count = 10;
					browseRequest.Offset = 0;
					browseRequest.FetchStoredFields = true;

					BooleanQuery resultQ = new BooleanQuery();

					TermQuery tq = new TermQuery(new Term(field, v1));
					tq.Boost = b1;
					resultQ.Add(tq, Occur.SHOULD);

					tq = new TermQuery(new Term(field, v2));
					tq.Boost = b2;
					resultQ.Add(tq, Occur.SHOULD);

					tq = new TermQuery(new Term(field, v3));
					tq.Boost = b3;
					resultQ.Add(tq, Occur.SHOULD);

					browseRequest.Query = resultQ;

					IBrowsable browser = new BoboBrowser(boboReader);
					result2 = browser.Browse(browseRequest);

					totalHits2 = result.NumHits;
				}


				// Check.
				Assert.AreEqual(totalHits, totalHits2);
				Assert.AreEqual(result.ToString(), result2.ToString());
			}
		}
	}
}
```

**Comment by NightOwl888** [comment]
Upon researching this, I discovered that there was a bug in the `ToString()` method in that it wasn't correctly returning the results of the facets, which will be addressed in the next release.

The results of the 2 queries are returning the exact same documents in the same order.

However, Lucene.Net uses a different scoring strategy than the default scoring of BoboBrowse.Net so the scores are different. If you look at the [Facet Scoring](https://linkedin.jira.com/wiki/display/BOBO/Facet+Query#FacetQuery-Scoring page) section, you can see that the score can be customized. So, I am not considering this a bug - it looks like this is normal behavior as the scores are defaulting to 1*boost and the score can be customized further if need be. Besides, scores are relative numbers, not absolute numbers and their relative values are correct. And there are quite a few tests covering this scenario that are all passing (TestFacetQueryBoost, TestFacetQuery, TestFacetQueryBoolean, TestFacetRangeQuery).

Comparing the output of `ToString()` is not a good way to determine if the 2 results are the same. In the Lucene.Net case there are no facets being returned and as previously mentioned the scores are different. The score could be fixed by building a custom scoring function, and the facets will be included if you add a `FacetSpec` as shown below, but you are better off comparing the individual values to ensure they appear in the same order and quantities.

```c#
// Bobo.
{
	BrowseRequest browseRequest = new BrowseRequest();
	browseRequest.Count = 10;
	browseRequest.Offset = 0;
	browseRequest.FetchStoredFields = true;

	BrowseSelection sel = new BrowseSelection(field);
	sel.Values = new string[] { v1, v2, v3 };
	browseRequest.AddSelection(sel);

	Dictionary<string, float> boostMap = new Dictionary<string, float>();
	boostMap.Add(v1, b1);
	boostMap.Add(v2, b2);
	boostMap.Add(v3, b3);

	// Add a custom scoring function factory that has the same scoring logic as Lucene.Net
	BoboBrowse.Net.Query.Scoring.IFacetTermScoringFunctionFactory scoringFunctionFactory = CustomScoringFunctionFactory();

	BoboBrowse.Net.Query.Scoring.FacetTermQuery resultQ = new BoboBrowse.Net.Query.Scoring.FacetTermQuery(sel, boostMap, scoringFunctionFactory);
	browseRequest.Query = resultQ;

	FacetSpec spec = new FacetSpec();
	spec.MaxCount = 10;
	browseRequest.SetFacetSpec(field, spec);

	IBrowsable browser = new BoboBrowser(boboReader);
	result = browser.Browse(browseRequest);

	totalHits = result.NumHits;
}

// Lucene.
{
	BrowseRequest browseRequest = new BrowseRequest();
	browseRequest.Count = 10;
	browseRequest.Offset = 0;
	browseRequest.FetchStoredFields = true;

	BooleanQuery resultQ = new BooleanQuery();

	TermQuery tq = new TermQuery(new Term(field, v1));
	tq.Boost = b1;
	resultQ.Add(tq, Occur.SHOULD);

	tq = new TermQuery(new Term(field, v2));
	tq.Boost = b2;
	resultQ.Add(tq, Occur.SHOULD);

	tq = new TermQuery(new Term(field, v3));
	tq.Boost = b3;
	resultQ.Add(tq, Occur.SHOULD);

	browseRequest.Query = resultQ;

        // If you includ a FacetSpec, the facets will be ruturned
	FacetSpec spec = new FacetSpec();
	spec.MaxCount = 10;
	browseRequest.SetFacetSpec(field, spec);

	IBrowsable browser = new BoboBrowser(boboReader);
	result2 = browser.Browse(browseRequest);

	totalHits2 = result.NumHits;
}
```

**Comment by NightOwl888** [comment]
Closing this issue as a wontfix. As I mentioned, ToString should not be relied upon for testing purposes. It is there for informational/debugging purposes only. That said, there were several ToString fixes in the latest release, [3.2.2](https://github.com/NightOwl888/BoboBrowse.Net/releases/tag/v3.2.2).
